module/userExtension/helper.js

```diff
@@ -211,304 +212,373 @@ module.exports = class UserExtensionHelper {
+                    continue outerloop;
+                  }
 
-                if (existingUser.roles && existingUser.roles.length > 0) {
-                  userRoleToUpdate = _.findIndex(existingUser.roles, {
-                    code: userRole.role,
-                  });
-                }
+                  const roles = platform_role_array;
+                  let entry = userInformation.platformRoles.find(
+                    (pr) => pr.programId.toString() === programObjectId.toString()
+                  );
 
-                if (!(userRoleToUpdate >= 0)) {
-                  userRoleToUpdate = existingUser.roles.length;
-                  userRoleMap[userRole.role].entities = new Array();
-                  existingUser.roles.push(userRoleMap[userRole.role]);
-                }
+                  if (!entry) {
+                    entry = { programId: programObjectId, roles: [] };
+                    userInformation.platformRoles.push(entry);
+                  }
 
-                existingUser.roles[userRoleToUpdate].entities = existingUser.roles[userRoleToUpdate].entities.map(
-                  (eachEntity) => eachEntity.toString(),
-                );
-
-                if (userRole.entityOperation == 'OVERRIDE') {
-                  existingUser.roles[userRoleToUpdate].entities = [userRole.entity];
-                  existingUser.roles[userRoleToUpdate].acl = userRole.acl;
-                } else if (userRole.entityOperation == 'APPEND' || userRole.entityOperation == 'ADD') {
-                  existingUser.roles[userRoleToUpdate].entities.push(userRole.entity);
-                  existingUser.roles[userRoleToUpdate].entities = _.uniq(existingUser.roles[userRoleToUpdate].entities);
-                } else if (userRole.entityOperation == 'REMOVE') {
-                  _.pull(existingUser.roles[userRoleToUpdate].entities, userRole.entity);
-                  removeUserFromEntity = true;
+                  for (const role of roles) {
+                    if (!entry.roles.includes(role)) {
+                      entry.roles.push(role);
+                      // Emit create event for new role
+                      kafkaEventPayloads.push({
+                        userId: userProfile.id,
+                        username: userProfile.username,
+                        programId: programObjectId,
+                        role,
+                        eventType: 'create',
+                      });
+                    }
+                  }
                 }
-
-                existingUser.roles[userRoleToUpdate].entities = existingUser.roles[userRoleToUpdate].entities.map(
-                  (eachEntity) => ObjectId(eachEntity),
-                );
-                updateQuery['roles'] = existingUser.roles;
               }
 
-              if (userRole.programOperation) {
-                let userPlatformRoleToUpdate;
+              user = await database.models.userExtension.create(userInformation);
```

Comment:
@borkarsaish65 Why are we directly calling db creation/delete/update from helper? Don't we have a directory called databaseQueries and from there we channel all db interactions right? Please check this once.

Reply:
done